### PR TITLE
Additional diagnostics for certificate errors

### LINKED ISSUE #178: Improve diagnostics when certificate is rejected
To improve diagnostics when a certificate is rejected for listener-accepted connection:
1. Log the error code returned from X509_STORE_CTX_get_error
2. Call AcceptConnection passing through preverified as a parameter, the application then has a chance to log the fact and optionally accept anyway

### PR DESCRIPTION
**Pull request by sidhoda**
Fixes #178

### CODE REVIEW

File: cpp/libs/src/asiodnp3/tls/MasterTLSServer.cpp [hunk 1]
```diff
@@ -71,28 +71,32 @@ bool MasterTLSServer::AcceptConnection(uint64_t sessionid, const asio::ip::tcp::
 bool MasterTLSServer::VerifyCallback(uint64_t sessionid, bool preverified, asio::ssl::verify_context& ctx)
 {
 	int depth = X509_STORE_CTX_get_error_depth(ctx.native_handle());
+	
+	// lookup the subject name
+	X509* cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+	char subjectName[512];
+	X509_NAME_oneline(X509_get_subject_name(cert), subjectName, 512);
```
Comment (sidhoda) [inline cpp/libs/src/asiodnp3/tls/MasterTLSServer.cpp]: Should the X509_NAME returned from X509_get_subject_name be freed? The OpenSSL docs suggest it shouldn't:

> X509_get_subject_name() returns the subject name of certificate x. The returned value is an internal pointer which MUST NOT be freed.

... however I've seen a few projects that do free it:

https://github.com/nginx/nginx/blob/3c44339bfed78c9ad3549edf2b7408c15d3ae51c/src/event/ngx_event_openssl.c#L782

**Comment by jadamcrain** [comment]
I'd trust the openssl docs over some random project. The _best_ way to know for certain would be to run the application under Valgrind and make sure all the heap allocations are certain.

**Comment by sidhoda** [comment]
Ignore my comment, I was looking at the wrong variable. It's actually freeing the char\* in that project. Sorry!

**Comment by jadamcrain** [comment]
Looks good.

**Pull request merged**